hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -77,6 +77,28 @@ object HoodieSparkSqlWriter {
   private var asyncCompactionTriggerFnDefined: Boolean = false
   private var asyncClusteringTriggerFnDefined: Boolean = false
 
+  def changeOperationToInsertIfRequired(writeOperationType: WriteOperationType, hoodieConfig: HoodieConfig)
```

Comment:
afaik in this case when changing op from upsert to insert one should also make sure that insert mode (SQL_INSERT_MODE) is set to "upsert", otherwise either duplicate record will be created (non-strict mode) or error will be thrown (strict mode)

Reply:
@kazdy : Would you be able to help us w/ a contribution on this. If you take this branch and work on top of it. As of this patch, we have tested auto generation of record keys works for spark-sql (CTAS and create table + inserts ). So, would appreciate if you can put up a fix on the strict/non-strict mode.